pkg/logql/log/dedup_filter_test.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	data = &testData{
+		Dev:  newLineLabelMap(devLabels, 5),
+		QA:   newLineLabelMap(qaLabels, 2),
+		Prod: newLineLabelMap(prodLabels, 5),
+	}
+
+	entries = data.allEntries()
+)
+
+func Test_Deduplication(t *testing.T) {
+
+	tests := []struct {
+		name         string
+		wantLines    [][]byte
+		labelFilters []string
+		inverted     bool
+	}{
+		{
+			name:         "dedup by (namespace)",
+			labelFilters: []string{"namespace"},
+			inverted:     false,
+			wantLines: [][]byte{
+				data.Dev.first(),  // first for devLabels (qaLabels has same "namespace")
+				data.Prod.first(), // first for prodLabels
+			},
+		},
+		{
+			name:         "dedup by (role)",
+			labelFilters: []string{"role"},
+			inverted:     false,
+			wantLines: append(
+				[][]byte{data.Dev.first(), data.QA.first()}, // only first entries with different "role" label
+				data.Prod.all()...), // include all of prodLabels since it has no "role" label
+		},
+		{
+			name:         "dedup without (role)",
+			labelFilters: []string{"role"},
+			inverted:     true,
+			// dedup without (x) will dedup by all labels that are not "x",
+			// so this will dedup by "namespace" and "cluster"
+			wantLines: [][]byte{data.Dev.first(), data.Prod.first()},
+		},
+		{
+			name:         "dedup without (unknown)",
+			labelFilters: []string{"unknown"},
+			inverted:     true,
+			// dedup without a label missing to all entries will result in a dedup by all present labels
+			wantLines: [][]byte{data.Dev.first(), data.QA.first(), data.Prod.first()},
+		},
+		{
+			name:         "dedup by (unknown)",
+			labelFilters: []string{"unknown"},
+			inverted:     false,
+			// dedup by a label missing to all entries is effectively a noop
+			wantLines: data.allLines(),
+		},
+		{
+			name:         "dedup by (cluster, namespace)",
+			labelFilters: []string{"cluster", "namespace"},
+			inverted:     false,
+			// dedup by multiple labels
+			wantLines: [][]byte{data.Dev.first(), data.Prod.first()},
+		},
+		{
+			name:         "dedup without (role, namespace)",
+			labelFilters: []string{"role", "namespace"},
+			inverted:     true,
+			// dedup without multiple labels
+			// - all lines have the same "cluster" label so only the first is returned
+			wantLines: [][]byte{data.Dev.first()},
+		},
+		{
+			name:         "dedup by ()",
+			labelFilters: nil,
+			inverted:     false,
+			// dedup without labels is effectively a noop
+			wantLines: data.allLines(),
+		},
+		{
+			name:         "dedup without ()",
+			labelFilters: nil,
+			inverted:     true,
+			// dedup without labels will result in a dedup by all present labels
+			wantLines: [][]byte{data.Dev.first(), data.QA.first(), data.Prod.first()},
+		},
+	}
+
+	for _, tt := range tests {
+		d := NewLineDedupFilter(tt.labelFilters, tt.inverted)
+
+		t.Run(tt.name, func(t *testing.T) {
+			var outLines [][]byte
+
+			for _, entry := range entries {
+				b := NewBaseLabelsBuilder().ForLabels(*entry.lbs, entry.lbs.Hash())
+				b.Reset()
```

Comment:
```suggestion
```

The `Reset()` call can be removed as it has just been freshly instantiated.

Reply:
Interesting, ok - I copied some existing code (can't recall which but here's an example: https://github.com/grafana/loki/blob/master/pkg/logql/log/fmt_test.go#L156) which does the same. Perhaps some reason for this? I can't see why though.
In any case, I'm removed them as suggested.

Reply:
Yeah I'm not sure the reset is required, my bad.